python/tank/descriptor/io_descriptor/downloadable.py

```diff
@@ -0,0 +1,139 @@
+# Copyright (c) 2016 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import os
+import uuid
+
+from .base import IODescriptorBase
+from ..errors import TankDescriptorError, TankError
+from ...util import filesystem
+
+from ... import LogManager
+log = LogManager.get_logger(__name__)
+
+
+class IODescriptorDownloadable(IODescriptorBase):
+    """
+    Base class for descriptors that can be downloaded locally to a path on disk.
+
+    In order to create a Descriptor associated with data that can
+    be downloaded locally to disk, it is necessary to derive from this class.
+    By default, the AppStore, Git and Shotgun entity descriptors can be downloaded
+    to disk and hence are of type :class: `IODescriptorDownloadable`.
+
+    Descriptor data can be downloaded by invoking the :meth: `download_local` on instances
+    of such derived classes. These classes are also expected to implement the
+    :meth: `_download_local` and optionally, the :meth: `_post_download`.
+
+    A general implementation of such a Descriptor class will be of the form:
+
+    eg. class MyNewDownloadableDescriptor(IODescriptorDownloadable):
+            def _download_local(self, destination_path):
+                # .. code to download data to destination_path
+
+            def _post_download(self, download_path):
+                # .. code that will be executed post download.
+    """
+    def download_local(self):
+        """
+        Downloads the data represented by the descriptor into the primary bundle
+        cache path.
+
+        It does so in a two step process. First, by downloading it to
+        a temporary bundle cache path (typically in a 'tmp/<uuid>' directory
+        in the bundle cache path), then, by moving the data to the primary bundle
+        cache path for that descriptor. This helps to guard against multiple
+        processes attempting to download the same descriptor simultaneously.
+        """
+
+        # Return if the descriptor exists locally.
+        if self.exists_local():
+            return
+
+        # cache into a temporary location
+        temporary_path = self._get_temporary_cache_path()
+
+        # move into primary location
+        target = self._get_primary_cache_path()
+
+        # ensure that the parent directory of the target is present.
+        # make sure we guard against multiple processes attempting to create it simultaneously.
+        try:
```

Comment:
I really like this implementation. It looks rock solid and makes the derived classes much simpler. :D